app/frontend/terrier/parts/select-field-part.ts

```diff
@@ -0,0 +1,214 @@
+import {SelectOption, SelectOptions} from "tuff-core/forms"
+import ContentPart from "./content-part"
+import Messages, { TypedKey } from "tuff-core/messages"
+import {PartTag} from "tuff-core/parts"
+import {Dropdown} from "../dropdowns"
+import Overlays, {AnchorResult} from "../overlays"
+
+/**
+ * Replacement part for a 'select' field that retrieves its options from a centralized data source
+ * rather than appending them to the DOM. When there are multiple select fields on the same page
+ * with the same set of options, this is more efficient than the standard 'select' field.
+ * @param options the options to add
+ * @param value the currently selected option value
+ * @param select_key the
+ */
+
+export type SelectFieldState = {
+    options: SelectOptions
+    selected_option: SelectOption
+}
+
+export class SelectFieldPart<T extends SelectFieldState> extends ContentPart<T> {
+    _toggleDropdownKey = Messages.untypedKey()
+
+    selectedOptionKey = Messages.typedKey<{ selected_option: SelectOption }>()
+
+    async init() {
+        await super.init()
+        this.onClick(this._toggleDropdownKey, () => {
+            this.toggleDropdown(SelectOptionsDropdown, {
+                options: this.state.options,
+                selected_option: this.state.selected_option,
+                select_key: this.selectedOptionKey
+            }, this.element)
+        })
+        this.listenMessage(this.selectedOptionKey, m => {
+            console.log('heard option selected')
+            this.onOptionSelected(m.data.selected_option)
+        }, { attach: "passive" })
+    }
+
+    get parentClasses(): Array<string> {
+        return ['tt-select-field', ...super.parentClasses]
+    }
+
+    onOptionSelected(selectedOption: SelectOption) {
+        this.state.selected_option = selectedOption
+        this.dirty()
```

Comment:
For future proofness, I would recommend using `assignState` here (`mergeState`, once it's published):

```suggestion
        this.assignState({ ...this.state, selected_option: selectedOption })
```

assignState should handle dirtying the part on its own.